torchmultimodal/utils/common.py

```diff
@@ -183,3 +185,34 @@ def to_tuple_tuple(
     if isinstance(param, tuple):
         param_fixed = (param,) * num_tuple
     return param_fixed
+
+
+def checkpoint_wrapper(fn: Callable) -> Callable:
+    """Decorator to render an nn.Module instance method in checkpointing mode to save memory for training"""
+
+    def inner(cls: nn.Module, *inputs: Any, **kwargs: Any) -> Tensor:
+        if cls.training:
+            # By default the checkpoint API stashes and restores the RNG state during each checkpointed
+            # segment such that checkpointed passes making use of RNG (e.g., through dropout, batch norm)
+            # have deterministic outputs as compared to non-checkpointed passes. This can incur a moderate
+            # performance hit which we mitigate by checkpointing either before and after the layer that
+            # requires RNG.
+            if "use_cache" in kwargs and kwargs["use_cache"] is True:
+                warnings.warn(
+                    "Using `cache` is incompatible with gradient checkpointing. Setting `use_cache=False`..."
+                )
+                kwargs["use_cache"] = False
+
+            def create_custom_forward(fn: Callable) -> Callable:
+                # Specifies what should the checkpoint API run in forward pass
+                def custom_forward(*inputs: Any) -> Callable:
+                    return fn(cls, *inputs, **kwargs)
+
+                return custom_forward
+
+            return checkpoint(create_custom_forward(fn), *inputs)
```

Comment:
any way to simplify three levels of nested functions? can this just be `checkpoint(fn(cls, *inputs, **kwargs))`? or do you need to format a specific way for `checkpoint`?

Reply:
We need to wrap `checkpoint` around `fn(**kwargs)` and apply it to `*inputs` where `kwargs` could be constants (such as booleans) and `inputs` are what we actually re-compute against during the backward pass. 
So it'd be like:
```
checkpoint(fn(use_cache), x, mask)
```

`cls` is for us to access the training status `cls.training`. It's not consumed by `checkpoint`.

So the three levels are:
1.  branch by training mode, handle `use_cache`
2.  tell `checkpoint` what to compute from
Maybe a better interpretation from PT core [here](https://pytorch.org/docs/stable/checkpoint.html):

> function – describes what to run in the forward pass of the model or part of the model.

3.  the actual forward logic

Reply:
ok I see. but it just seems like `create_custom_forward` and `custom_forward` achieve the same thing as just `checkpoint(fn(cls, *inputs, **kwargs), *inputs)`. Isn't that what `create_custom_forward` is returning anyway?

Reply:
`create_custom_forward` is trying to tease out the boolean arguments (or whatever provided as `**kwargs`) so that we only forward pass on the tensor arguments (see my additional comment below). 

```
def create_custom_forward(fn):
                # Specifies what should the checkpoint API run in forward pass
                def custom_forward(*inputs):    # custom_forward only interfaces *inputs
                    return fn(cls, *inputs, **kwargs)

                return custom_forward
```

But I am trying to validate if this separation is necessary and will update once I can confirm it.

Reply:
So yeah I "asked a pytorch developer" and both yours and mine approaches are ok. For simplicity, I will amend this either in-place or a follow-up diff.

Reply:
cool, thanks for making that post about it

Reply:
Unfortunately, I am not able to make the simplification as `checkpoint` does not accept kwargs other than those defined in their API. So I had to create a custom function to hide kwargs such as `use_cache` or `attn_mask` from `checkpoint`.